tool/generator/bcd.dart

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert' hide json;
+import 'dart:js_interop';
+
+import 'package:path/path.dart' as p;
+
+import 'filesystem_api.dart';
+
+/// A class to read from the browser-compat-data files and parse interface and
+/// property status (standards track, experimental, deprecated) and supported
+/// browser (chrome, safari, firefox) info.
+class BrowserCompatData {
+  static BrowserCompatData read() {
+    final path =
+        p.join('node_modules', '@mdn', 'browser-compat-data', 'data.json');
+    final content = (fs.readFileSync(
+      path.toJS,
+      JSReadFileOptions(encoding: 'utf8'.toJS),
+    ) as JSString)
+        .toDart;
+
+    final api = (jsonDecode(content) as Map)['api'] as Map<String, dynamic>;
+    final interfaces = api.keys
+        .where((key) => !key.startsWith('_'))
+        .map((key) => BCDInterfaceStatus(key, api[key] as Map<String, dynamic>))
+        .toList();
+    return BrowserCompatData(Map.fromIterable(
+      interfaces,
+      key: (i) => (i as BCDInterfaceStatus).name,
+    ));
+  }
+
+  final Map<String, BCDInterfaceStatus> interfaces;
+
+  BrowserCompatData(this.interfaces);
+
+  BCDInterfaceStatus? interfaceFor(String name) => interfaces[name];
+}
+
+class BCDInterfaceStatus extends BCDItem {
+  late final Map<String, BCDPropertyStatus> properties;
+
+  BCDInterfaceStatus(super.name, super.json) {
+    final names = json.keys.where((key) => !key.startsWith('_'));
+    properties = Map.fromIterable(
+      names,
+      value: (key) => BCDPropertyStatus(
+          key as String, json[key] as Map<String, dynamic>, this),
```

Comment:
The MDN exposes compatibility info for dictionary interfaces with an `_option` suffix e.g. https://github.com/mdn/browser-compat-data/blob/cfa15e085ceb88cdee391ae09bd52ee3674d5ecc/api/MediaDevices.json#L264C10-L264C29. We could handle this here by having a separate "dictionaryOptions" field in `BCDInterfaceStatus` or we can just do the lookup in the generation script.

Reply:
Just to help me think about the organization in the bcd.json file, there are ~16 `foo_option` instances in there (all fields off of interface properties?).

https://gist.github.com/devoncarew/81c987aa55637b6014c5b400a7edf022

Reply:
Funnily enough, that info is actually a separate thing altogether and I think they're suffixed with `_parameter`: https://github.com/mdn/browser-compat-data/blob/c2761cb84874ff5b9b0a230e7a557a5a43b9b939/api/EventTarget.json#L160C10-L160C27. :D Also see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#browser_compatibility.

This is for what options can go in an options object that is passed to an API. https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getDisplayMedia#browser_compatibility

There's also "_parameter_optional" apparently for optional parameters: https://github.com/mdn/browser-compat-data/blob/c2761cb84874ff5b9b0a230e7a557a5a43b9b939/api/HTMLTableRowElement.json#L378. 

I think we may want to validate all the strings and check for `_` to make sure we're handling the various types of compatibility info. There's also some guidance here: https://github.com/mdn/browser-compat-data/blob/main/docs/data-guidelines/index.md#parameters-and-parameter-object-features, but it doesn't seem to catch every variant unfortunately. Realistically, there are probably not enough APIs that use all the variants for this to matter from our perspective, but something to watch out for.

Reply:
OK, taking another pass through this, it looks like neither `_option` nor `_parameter` show up in the first two levels of the data (the interfaces and their properties).

Separately, I do see some types we're generating interop code for that aren't represented in the BCD data. I expect we'll need to adjust how we parse and use the data as we start to filter how much of the IDL we generate code for.

Reply:
> OK, taking another pass through this, it looks like neither _option nor _parameter show up in the first two levels of the data (the interfaces and their properties).

Right, we can do this in another CL to grab more information about a specific member or constructor.

> Separately, I do see some types we're generating interop code for that aren't represented in the BCD data. I expect we'll need to adjust how we parse and use the data as we start to filter how much of the IDL we generate code for.

I suspect the majority of these cases to be because the types are new and are still being drafted, and therefore MDN doesn't have information on them. It'll be useful to validate that, but I suspect if the type is not present, we can safely omit the type.